### ISSUE TITLE
Performance / Memory Usage

### ISSUE THREAD

**Issue by mstaack** [zbateson/MailMimeParser#4]
hey!
thanks for this package..it works perfectly!

i compared the memory usage and time needed to parse ~10 messages
and saw it uses way more  than the mailparse extension.

is there a way to optimize this?

cheers max

**Comment by zbateson** [comment]
Hi @mstaack 

Thanks for the feedback and for running that comparison :)

Unfortunately comparing to mailparse for performance would be a losing battle -- for one mailparse is a native extension for PHP, so it comes pre-compiled and wouldn't require the overhead of run-time interpretation like this does, etc...

Secondly, it provides only procedural functions and 'does less' for you... for instance mailparse_msg_get_part_data gives you an array of headers/values, but it's still up to you to parse the values further if necessary (like email address headers into name/values, or list names, or list names with name values, etc... etc...) whereas with MailMimeParser you get objects for everything, but of course with a performance cost.

In conclusion I don't think we'd ever get close to mailparse or imap_* performance wise, but I'd certainly be happy to get it performing better where possible.

I'd be very interested in learning what the difference you saw was between the two as well, and perhaps some metrics for comparison (average/min/max and 10 vs 50 vs 100) if you have the time :)

**Comment by zbateson** [comment]
One idea occurred to me -- for many cases the overhead of reading and parsing the message body and it's parts may be unnecessary and reading the headers only would be enough.  Reading the message itself could be delayed till part of the body is requested by a client.

Another thing about memory usage... currently the implementation copies the passed stream to a php://temp stream potentially copying the entire stream to memory (php is at liberty to write to a temp file or keep in memory).  The idea is to close the underlying stream as quickly as possible, but perhaps that's unnecessary and the library is trying to do too much (it has an "isSmtp" parameter in the parse method of MailMimeParser.)

**Comment by mstaack** [comment]
@zbateson thanks for your fast response!

i just used this quick&dirty script:
https://gist.github.com/mstaack/9f77a98661e8438ef365

dont have much time right now, but it was using around 4x more memory

**Comment by zbateson** [comment]
Cool, thanks @mstaack that is very much appreciated!

I'll dig around, it is a pretty big difference.  My main concern would be with the difference in performance (150ms vs 1.2s)  I'll create a similar script to compare them against my test emails to avoid any overhead/uncertainty with using an open connection and make the tests more comparable, etc... when I have some time also and see where the bottlenecks may be :)

It may indeed be because I'm parsing the entire email and in cases like this, just reading the headers would be enough.

**Comment by mstaack** [comment]
I can highly suggest to use PHP blackfire! Its just pretty awesome in these cases! Give it try!

Thanks for your time

**Comment by zbateson** [comment]
Looks interesting, I'll look into it :)

**Comment by mstaack** [comment]
@zbateson your up on any chat environment like slack/icq/whatever would love to have some tech talks ;)

**Comment by zbateson** [comment]
@mstaack - I'm a bit sporadic with my chat availability, but I'll message you my gmail.  You can occasionally find me on gchat or I'm pretty responsive to emails :)

**Comment by zbateson** [comment]
My implementation of copyToTmpFile was really bad.  I pushed a fix for it to dev-master, but it only addresses streams with large amounts of data in them (parsing m3005.txt 55 times under my tests/_data dir was taking 12s for me, and is now down to 4s with this).  It's still significantly slower than mailparse, so I'll keep investigating as well.

For smaller files, e.g. running all 55 test files is a difference of 29ms for mailparse and 256ms for MailMimeParser.  Almost 10x slower, but not as concerning to me.

Memory usage is lower for mailparse when running through all 55 test files (~700KB).  Running mailparse 55 times with my largest file (m3005.txt again), it ends up using more memory than MailMimeParser (5.25MB).  Mine consistently only seems to use 3.75MB for me regardless of what files are put through it.  With xdebug enabled it goes up to 4MB, but I'm not seeing anywhere close to the 15MB you're getting.  Overall I'm still not really concerned about the memory usage at this point, it should be okay.

**Comment by mstaack** [comment]
@zbateson have you made any progress reducing time/memory usage? considering switchting to your solution and dropping https://github.com/php-mime-mail-parser/php-mime-mail-parser..

using this in a really high traffic site.. :)

**Comment by zbateson** [comment]
Hi @mstaack --

Beyond my fix for copyToTmpFile, most of my work has been with making the library more reliable.  I do have a future plan to make parsing headers an on-demand process rather than reading and parsing all headers right away which may improve things a bit... but it hasn't been my focus at this point.

I have had some pretty good feedback of others using it with at least large files, and possibly high traffic.  I think @postme is using it pretty extensively with a pretty high load, but perhaps he can expand on that :)

A big plus with my library is it's usage of streams to avoid storing the entire content of an email in memory to parse it... for instance my mailparse test with a large file versus my MailMimeParser test, it seems mine may be handling things a bit better.  A user on StackOverflow reported he's using it with large files and no issues as well:

> Works great! I stream 40+MB emails from an AWS SES inbox on S3 with zero problems. Excellent library. – bishop 

http://stackoverflow.com/a/33948593/335059

I'd certainly be happy to get it working better, but would also appreciate help finding bottlenecks as well :) if you do decide to switch over it'll be nice to know the impact and ideas on where we can improve MMP.  I do still have to stress the difference between using the mailparse extension functions versus my library as well, since mailparse functions (which php-mime-mail-parser relies on) are compiled and built-in, whereas mine is all php code.

All the best!
Zaahid

**Comment by mstaack** [comment]
Thanks for your information! 

@postme how do you actuall fetch messages via imap from php? native extension?

Cheers max

**Comment by postme** [comment]
Hi @mstaack I use the same method as referred to in the link above, i.e. e-mail comes in to SES and is stored in an S3 bucket where it triggers a Lambda for processing. The Lambda (actually a series of Lambdas) is built with nodejs (bootstrap) and PHP using @zbateson MailMimeParser library. There is no IMAP or PHP extension involved. This allows for an unlimited amount of email processing without having to deal with processor or memory constraints. My system can scale from hundreds of messages to millions of messages within a second with this architecture.
